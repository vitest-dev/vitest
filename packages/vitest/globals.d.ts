declare global {
  let suite: typeof import('vitest')['suite']
  let test: typeof import('vitest')['test']
  let chai: typeof import("vitest")["chai"]
  let describe: typeof import('vitest')['describe']
  let it: typeof import('vitest')['it']
  let expectTypeOf: typeof import('vitest')['expectTypeOf']
  let assertType: typeof import('vitest')['assertType']
  let expect: typeof import('vitest')['expect']
  let assert: typeof import('vitest')['assert']
  let vitest: typeof import('vitest')['vitest']
  let vi: typeof import('vitest')['vitest']
  let beforeAll: typeof import('vitest')['beforeAll']
  let afterAll: typeof import('vitest')['afterAll']
  let beforeEach: typeof import('vitest')['beforeEach']
  let afterEach: typeof import('vitest')['afterEach']
  let onTestFailed: typeof import('vitest')['onTestFailed']
  let onTestFinished: typeof import('vitest')['onTestFinished']
}
export {}
