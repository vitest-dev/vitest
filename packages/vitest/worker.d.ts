export * from './dist/worker.js'
