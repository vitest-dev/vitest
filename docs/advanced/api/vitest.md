---
outline: deep
title: Vitest API
---

# Vitest

Vitest instance requires the current test mode. It can be either:

- `test` when running runtime tests
- `benchmark` when running benchmarks <Badge type="warning">experimental</Badge>

::: details New in Vitest 3
Vitest 3 is one step closer to stabilising the public API. To achieve that, we deprecated and removed some of the previously public methods on the `Vitest` class. These APIs were made private:

- `configOverride` (use [`setGlobalTestNamePattern`](#setglobaltestnamepattern) or [`enableSnapshotUpdate`](#enablesnapshotupdate))
- `coverageProvider`
- `filenamePattern`
- `runningPromise`
- `closingPromise`
- `isCancelling`
- `coreWorkspaceProject`
- `resolvedProjects`
- `_browserLastPort`
- `_options`
- `reporters`
- `vitenode`
- `runner`
- `pool`
- `setServer`
- `_initBrowserServers`
- `rerunTask`
- `changeProjectName`
- `changeNamePattern`
- `changeFilenamePattern`
- `rerunFailed`
- `_createRootProject` (renamed to `_ensureRootProject`, but still private)
- `filterTestsBySource` (this was moved to the new internal `vitest.specifications` instance)
- `runFiles` (use [`runTestSpecifications`](#runtestspecifications) instead)
- `onAfterSetServer`

These APIs were deprecated:
- `invalidates`
- `changedTests` (use [`onFilterWatchedSpecification`](#onfilterwatchedspecification) instead)
- `server` (use [`vite`](#vite) instead)
- `getProjectsByTestFile` (use [`getModuleSpecifications`](#getmodulespecifications) instead)
- `getFileWorkspaceSpecs` (use [`getModuleSpecifications`](#getmodulespecifications) instead)
- `getModuleProjects` (filter by [`this.projects`](#projects) yourself)
- `updateLastChanged` (renamed to [`invalidateFile`](#invalidatefile))
- `globTestSpecs` (use [`globTestSpecifications`](#globtestspecifications) instead)
- `globTestFiles` (use [`globTestSpecifications`](#globtestspecifications) instead)
- `listFile` (use [`getRelevantTestSpecifications`](#getrelevanttestspecifications) instead)
:::

## mode

### test

Test mode will only call functions inside `test` or `it`, and throws an error when `bench` is encountered. This mode uses `include` and `exclude` options in the config to find test files.

### benchmark <Badge type="warning">experimental</Badge>

Benchmark mode calls `bench` functions and throws an error, when it encounters `test` or `it`. This mode uses `benchmark.include` and `benchmark.exclude` options in the config to find benchmark files.

## config

The root (or global) config. If workspace feature is enabled, projects will reference this as `globalConfig`.

::: warning
This is Vitest config, it doesn't extend _Vite_ config. It only has resolved values from the `test` property.
:::

## vite

This is a global [`ViteDevServer`](https://vite.dev/guide/api-javascript#vitedevserver).

## state <Badge type="warning">experimental</Badge>

::: warning
Public `state` is an experimental API (except `vitest.state.getReportedEntity`). Breaking changes might not follow SemVer, please pin Vitest's version when using it.
:::

Global state stores information about the current tests. It uses the same API from `@vitest/runner` by default, but we recommend using the [Reported Tasks API](/advanced/reporters#reported-tasks) instead by calling `state.getReportedEntity()` on the `@vitest/runner` API:

```ts
const task = vitest.state.idMap.get(taskId) // old API
const testCase = vitest.state.getReportedEntity(task) // new API
```

In the future, the old API won't be exposed anymore.

## snapshot

The global snapshot manager. Vitest keeps track of all snapshots using the `snapshot.add` method.

You can get the latest summary of snapshots via the `vitest.snapshot.summay` property.

## cache

Cache manager that stores information about latest test results and test file stats. In Vitest itself this is only used by the default sequencer to sort tests.

## projects

An array of [test projects](/advanced/api/test-project) that belong to the user's workspace. If the user did not specify a custom workspace, the workspace will only have a [root project](#getrootproject).

Vitest will ensure that there is always at least one project in the workspace. If the user specifies a non-existent `--project` name, Vitest will throw an error.

## getRootProject

```ts
function getRootProject(): TestProject
```

This returns the root test project. The root project generally doesn't run any tests and is not included in `vitest.projects` unless the user explicitly includes the root config in their workspace, or the workspace is not defined at all.

The primary goal of the root project is to setup the global config. In fact, `rootProject.config` references `rootProject.globalConfig` and `vitest.config` directly:

```ts
rootProject.config === rootProject.globalConfig === rootProject.vitest.config
```

## provide

```ts
function provide<T extends keyof ProvidedContext & string>(
  key: T,
  value: ProvidedContext[T],
): void
```

Vitest exposes `provide` method which is a shorthand for `vitest.getRootProject().provide`. With this method you can pass down values from the main thread to tests. All values are checked with `structuredClone` before they are stored, but the values themselves are not cloned.

To recieve the values in the test, you need to import `inject` method from `vitest` entrypont:

```ts
import { inject } from 'vitest'
const port = inject('wsPort') // 3000
```

For better type safety, we encourage you to augment the type of `ProvidedContext`:

```ts
import { createVitest } from 'vitest/node'

const vitest = await createVitest('test', {
  watch: false,
})
vitest.provide('wsPort', 3000)

declare module 'vitest' {
  export interface ProvidedContext {
    wsPort: number
  }
}
```

::: warning
Technically, `provide` is a method of [`TestProject`](/advanced/api/test-project), so it is limited to the specific project. However, all projects inherit the values from the core project which makes `vitest.provide` universal way of passing down values to tests.
:::

## getProvidedContext

```ts
function getProvidedContext(): ProvidedContext
```

This returns the root context object. This is a shorthand for `vitest.getRootProject().getProvidedContext`.

## getProjectByName

```ts
function getProjectByName(name: string): TestProject
```

This method returns the project by its name. Simillar to calling `vitest.projects.find`.

::: warning
In case the project doesn't exist, this method will return the root project - make sure to check the names again if the project you are looking for is the one returned.

If user didn't customize a name, the Vitest will assign an empty string as a name.
:::

## globTestSpecifications

```ts
function globTestSpecifications(
  filters?: string[],
): Promise<TestSpecification[]>
```

This method constructs new [test specifications](/advanced/api/test-specification) by collecting every test in all projects with [`project.globTestFiles`](/advanced/api/test-project#globtestfiles). It accepts string filters to match the test files - these are the same filters that [CLI supports](/guide/filtering#cli).

This method automatically caches all test specifications. When you call [`getModuleSpecifications`](#getmodulespecifications) next time, it will return the same specifications unless [`clearSpecificationsCache`](#clearspecificationscache) was called before that.

::: warning
As of Vitest 3, it's possible to have multiple test specifications with the same module ID (file path) if `poolMatchGlob` has several pools or if `typecheck` is enabled. This possibility will be removed in Vitest 4.
:::

```ts
const specifications = await vitest.globTestSpecifications(['my-filter'])
// [TestSpecification{ moduleId: '/tests/my-filter.test.ts' }]
console.log(specifications)
```

## getRelevantTestSpecifications

```ts
function getRelevantTestSpecifications(
  filters?: string[]
): Promise<TestSpecification[]>
```

This method resolves every test specification by calling [`project.globTestFiles`](/advanced/api/test-project#globtestfiles). It accepts string filters to match the test files - these are the same filters that [CLI supports](/guide/filtering#cli). If `--changed` flag was specified, the list will be filtered to include only files that changed. `getRelevantTestSpecifications` doesn't run any test files.

::: warning
This method can be slow because it needs to filter `--changed` flags. Do not use it if you just need a list of test files.

- If you need to get the list of specifications for known test files, use [`getModuleSpecifications`](#getmodulespecifications) instead.
- If you need to get the list of all possible test files, use [`globTestSpecifications`](#globtestspecifications).
:::

## mergeReports

```ts
function mergeReports(directory?: string): Promise<TestRunResult>
```

Merge reports from multiple runs located in the specified directory (value from `--merge-reports` if not specified). This value can also be set on `config.mergeReports` (by default, it will read `.vitest-reports` folder).

Note that the `directory` will always be resolved relative to the working directory.

This method is called automatically by [`startVitest`](/advanced/guide/tests) if `config.mergeReports` is set.

## collect

```ts
function collect(filters?: string[]): Promise<TestRunResult>
```

Execute test files without running test callbacks. `collect` returns unhandled errors and an array of [test modules](/advanced/api/test-module). It accepts string filters to match the test files - these are the same filters that [CLI supports](/guide/filtering#cli).

This method resolves tests specifications based on the config `include`, `exclude`, and `includeSource` values. Read more at [`project.globTestFiles`](/advanced/api/test-project#globtestfiles). If `--changed` flag was specified, the list will be filtered to include only files that changed.

::: warning
Note that Vitest doesn't use static analysis to collect tests. Vitest will run every test file in isolation, just like it runs regular tests.

This makes this method very slow, unless you disable isolation before collecting tests.
:::

## start

```ts
function start(filters?: string[]): Promise<TestRunResult>
```

Initialize reporters, the coverage provider, and run tests. This method accepts string filters to match the test files - these are the same filters that [CLI supports](/guide/filtering#cli).

::: warning
This method should not be called if [`vitest.init()`](#init) is also invoked. Use [`runTestSpecifications`](#runtestspecifications) or [`rerunTestSpecifications`](#reruntestspecifications) instead if you need to run tests after Vitest was inititalised.
:::

This method is called automatically by [`startVitest`](/advanced/guide/tests) if `config.mergeReports` and `config.standalone` are not set.

## init

```ts
function init(): Promise<void>
```

Initialize reporters and the coverage provider. This method doesn't run any tests. If the `--watch` flag is provided, Vitest will still run changed tests even if this method was not called.

Internally, this method is called only if [`--standalone`](/guide/cli#standalone) flag is enabled.

::: warning
This method should not be called if [`vitest.start()`](#start) is also invoked.
:::

This method is called automatically by [`startVitest`](/advanced/guide/tests) if `config.standalone` is set.

## getModuleSpecifications

```ts
function getModuleSpecifications(moduleId: string): TestSpecification[]
```

Returns a list of test specifications related to the module ID. The ID should already be resolved to an absolute file path. If ID doesn't match `include` or `includeSource` patterns, the returned array will be empty.

This method can return already cached specifications based on the `moduleId` and `pool`. But note that [`project.createSpecification`](/advanced/api/test-project#createspecification) always returns a new instance and it's not cached automatically. However, specifications are automatically cached when [`runTestSpecifications`](#runtestspecifications) is called.

::: warning
As of Vitest 3, this method uses a cache to check if the file is a test. To make sure that the cache is not empty, call [`globTestSpecifications`](#globtestspecifications) at least once.
:::

## clearSpecificationsCache

```ts
function clearSpecificationsCache(moduleId?: string): void
```

Vitest automatically caches test specifications for each file when [`globTestSpecifications`](#globtestspecifications) or [`runTestSpecifications`](#runtestspecifications) is called. This method clears the cache for the given file or the whole cache alltogether depending on the first argument.

## runTestSpecifications

```ts
function runTestSpecifications(
  specifications: TestSpecification[],
  allTestsRun = false
): Promise<TestRunResult>
```

This method runs every test based on the received [specifications](/advanced/api/test-specification). The second argument, `allTestsRun`, is used by the coverage provider to determine if it needs to instrument coverage on _every_ file in the root (this only matters if coverage is enabled and `coverage.all` is set to `true`).

::: warning
This method doesn't trigger `onWatcherRerun`, `onWatcherStart` and `onTestsRerun` callbacks. If you are rerunning tests based on the file change, consider using [`rerunTestSpecifications`](#reruntestspecifications) instead.
:::

## rerunTestSpecifications

```ts
function rerunTestSpecifications(
  specifications: TestSpecification[],
  allTestsRun = false
): Promise<TestRunResult>
```

This method emits `reporter.onWatcherRerun` and `onTestsRerun` events, then it runs tests with [`runTestSpecifications`](#runtestspecifications). If there were no errors in the main process, it will emit `reporter.onWatcherStart` event.

## updateSnapshot

```ts
function updateSnapshot(files?: string[]): Promise<TestRunResult>
```

Update snapshots in specified files. If no files are provided, it will update files with failed tests and obsolete snapshots.

## collectTests

```ts
function collectTests(
  specifications: TestSpecification[]
): Promise<TestRunResult>
```

Execute test files without running test callbacks. `collectTests` returns unhandled errors and an array of [test modules](/advanced/api/test-module).

This method works exactly the same as [`collect`](#collect), but you need to provide test specifications yourself.

::: warning
Note that Vitest doesn't use static analysis to collect tests. Vitest will run every test file in isolation, just like it runs regular tests.

This makes this method very slow, unless you disable isolation before collecting tests.
:::

## cancelCurrentRun

```ts
function cancelCurrentRun(reason: CancelReason): Promise<void>
```

This method will gracefully cancel all ongoing tests. It will wait for started tests to finish running and will not run tests that were scheduled to run but haven't started yet.

## setGlobalTestNamePattern

```ts
function setGlobalTestNamePattern(pattern: string | RegExp): void
```

This methods overrides the global [test name pattern](/config/#testnamepattern).

::: warning
This method doesn't start running any tests. To run tests with updated pattern, call [`runTestSpecifications`](#runtestspecifications).
:::

## resetGlobalTestNamePattern

```ts
function resetGlobalTestNamePattern(): void
```

This methods resets the [test name pattern](/config/#testnamepattern). It means Vitest won't skip any tests now.

::: warning
This method doesn't start running any tests. To run tests without a pattern, call [`runTestSpecifications`](#runtestspecifications).
:::

## enableSnapshotUpdate

```ts
function enableSnapshotUpdate(): void
```

Enable the mode that allows updating snapshots when running tests. Every test that runs after this method is called will update snapshots. To disable the mode, call [`resetSnapshotUpdate`](#resetsnapshotupdate).

::: warning
This method doesn't start running any tests. To update snapshots, run tests with [`runTestSpecifications`](#runtestspecifications).
:::

## resetSnapshotUpdate

```ts
function resetSnapshotUpdate(): void
```

Disable the mode that allows updating snapshots when running tests. This method doesn't start running any tests.

## invalidateFile

```ts
function invalidateFile(filepath: string): void
```

This method invalidates the file in the cache of every project. It is mostly useful if you rely on your own watcher because Vite's cache persist in memory.

::: danger
If you disable Vitest's watcher but keep Vitest running, it is important to manually clear the cache with this method because there is no way to disable the cache. This method will also invalidate file's importers.
:::

## import

<!--@include: ./import-example.md-->

Import a file using Vite module runner. The file will be transformed by Vite with the global config and executed in a separate context. Note that `moduleId` will be relative to the `config.root`.

::: danger
`project.import` reuses Vite's module graph, so importing the same module using a regular import will return a different module:

```ts
import * as staticExample from './example.js'
const dynamicExample = await vitest.import('./example.js')

dynamicExample !== staticExample // ✅
```
:::

::: info
Internally, Vitest uses this method to import global setups, custom coverage providers, workspace file, and custom reporters, meaning all of them share the same module graph as long as they belong to the same Vite server.
:::

## close

```ts
function close(): Promise<void>
```

Closes all projects and their associated resources. This can only be called once; the closing promise is cached until the server restarts.

## exit

```ts
function exit(force = false): Promise<void>
```

Closes all projects and exit the process. If `force` is set to `true`, the process will exit immediately after closing the projects.

This method will also forcefuly call `process.exit()` if the process is still active after [`config.teardownTimeout`](/config/#teardowntimeout) milliseconds.

## shouldKeepServer

```ts
function shouldKeepServer(): boolean
```

This method will return `true` if the server should be kept running after the tests are done. This usually means that the `watch` mode was enabled.

## onServerRestart

```ts
function onServerRestart(fn: OnServerRestartHandler): void
```

Register a handler that will be called when the server is restarted due to a config change.

## onCancel

```ts
function onCancel(fn: (reason: CancelReason) => Awaitable<void>): void
```

Register a handler that will be called when the test run is cancelled with [`vitest.cancelCurrentRun`](#cancelcurrentrun).

## onClose

```ts
function onClose(fn: () => Awaitable<void>): void
```

Register a handler that will be called when the server is closed.

## onTestsRerun

```ts
function onTestsRerun(fn: OnTestsRerunHandler): void
```

Register a handler that will be called when the tests are rerunning. The tests can rerun when [`rerunTestSpecifications`](#reruntestspecifications) is called manually or when a file is changed and the built-in watcher schedules a rerun.

## onFilterWatchedSpecification

```ts
function onFilterWatchedSpecification(
  fn: (specification: TestSpecification) => boolean
): void
```
Register a handler that will be called when a file is changed. This callback should return `true` or `false`, indicating whether the test file needs to be rerun.

With this method, you can hook into the default watcher logic to delay or discard tests that the user doesn't want to keep track of at the moment:

```ts
const continuesTests: string[] = []

myCustomWrapper.onContinuesRunEnabled(testItem =>
  continuesTests.push(item.fsPath)
)

vitest.onFilterWatchedSpecification(specification =>
  continuesTests.includes(specification.moduleId)
)
```

Vitest can create different specifications for the same file depending on the `pool` or `locations` options, so do not rely on the reference. Vitest can also return cached specification from [`vitest.getModuleSpecifications`](#getmodulespecifications) - the cache is based on the `moduleId` and `pool`. Note that [`project.createSpecification`](/advanced/api/test-project#createspecification) always returns a new instance.
