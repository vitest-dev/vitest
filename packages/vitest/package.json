{
  "name": "vitest",
  "type": "module",
  "version": "4.0.0-beta.17",
  "description": "Next generation testing framework powered by Vite",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://opencollective.com/vitest",
  "homepage": "https://vitest.dev",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vitest-dev/vitest.git",
    "directory": "packages/vitest"
  },
  "bugs": {
    "url": "https://github.com/vitest-dev/vitest/issues"
  },
  "keywords": [
    "vite",
    "vitest",
    "test",
    "jest"
  ],
  "sideEffects": false,
  "imports": {
    "#module-evaluator": {
      "types": "./dist/module-evaluator.d.ts",
      "default": "./dist/module-evaluator.js"
    }
  },
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.js"
      },
      "require": {
        "types": "./index.d.cts",
        "default": "./index.cjs"
      }
    },
    "./browser": {
      "types": "./browser/context.d.ts",
      "default": "./browser/context.js"
    },
    "./package.json": "./package.json",
    "./optional-types.js": {
      "types": "./optional-types.d.ts"
    },
    "./src/*": "./src/*",
    "./globals": {
      "types": "./globals.d.ts"
    },
    "./jsdom": {
      "types": "./jsdom.d.ts"
    },
    "./importMeta": {
      "types": "./importMeta.d.ts"
    },
    "./import-meta": {
      "types": "./import-meta.d.ts"
    },
    "./node": {
      "types": "./dist/node.d.ts",
      "default": "./dist/node.js"
    },
    "./internal/browser": {
      "types": "./dist/browser.d.ts",
      "default": "./dist/browser.js"
    },
    "./internal/module-runner": {
      "types": "./dist/module-runner.d.ts",
      "default": "./dist/module-runner.js"
    },
    "./runners": {
      "types": "./dist/runners.d.ts",
      "default": "./dist/runners.js"
    },
    "./suite": {
      "types": "./dist/suite.d.ts",
      "default": "./dist/suite.js"
    },
    "./environments": {
      "types": "./dist/environments.d.ts",
      "default": "./dist/environments.js"
    },
    "./config": {
      "types": "./config.d.ts",
      "require": "./dist/config.cjs",
      "default": "./dist/config.js"
    },
    "./coverage": {
      "types": "./coverage.d.ts",
      "default": "./dist/coverage.js"
    },
    "./reporters": {
      "types": "./dist/reporters.d.ts",
      "default": "./dist/reporters.js"
    },
    "./snapshot": {
      "types": "./dist/snapshot.d.ts",
      "default": "./dist/snapshot.js"
    },
    "./mocker": {
      "types": "./dist/mocker.d.ts",
      "default": "./dist/mocker.js"
    }
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "bin": {
    "vitest": "./vitest.mjs"
  },
  "files": [
    "*.cjs",
    "*.d.cts",
    "*.d.ts",
    "*.mjs",
    "bin",
    "browser",
    "dist"
  ],
  "engines": {
    "node": "^20.0.0 || ^22.0.0 || >=24.0.0"
  },
  "scripts": {
    "build": "premove dist && rollup -c",
    "dev": "NODE_OPTIONS=\"--max-old-space-size=8192\" rollup -c --watch -m inline"
  },
  "peerDependencies": {
    "@edge-runtime/vm": "*",
    "@types/debug": "^4.1.12",
    "@types/node": "^20.0.0 || ^22.0.0 || >=24.0.0",
    "@vitest/browser-playwright": "workspace:*",
    "@vitest/browser-preview": "workspace:*",
    "@vitest/browser-webdriverio": "workspace:*",
    "@vitest/ui": "workspace:*",
    "happy-dom": "*",
    "jsdom": "*"
  },
  "peerDependenciesMeta": {
    "@edge-runtime/vm": {
      "optional": true
    },
    "@types/debug": {
      "optional": true
    },
    "@types/node": {
      "optional": true
    },
    "@vitest/browser-playwright": {
      "optional": true
    },
    "@vitest/browser-preview": {
      "optional": true
    },
    "@vitest/browser-webdriverio": {
      "optional": true
    },
    "@vitest/ui": {
      "optional": true
    },
    "happy-dom": {
      "optional": true
    },
    "jsdom": {
      "optional": true
    }
  },
  "dependencies": {
    "@vitest/expect": "workspace:*",
    "@vitest/mocker": "workspace:*",
    "@vitest/pretty-format": "workspace:*",
    "@vitest/runner": "workspace:*",
    "@vitest/snapshot": "workspace:*",
    "@vitest/spy": "workspace:*",
    "@vitest/utils": "workspace:*",
    "debug": "catalog:",
    "es-module-lexer": "^1.7.0",
    "expect-type": "^1.2.2",
    "magic-string": "catalog:",
    "pathe": "catalog:",
    "picomatch": "^4.0.3",
    "std-env": "catalog:",
    "tinybench": "^2.9.0",
    "tinyexec": "^0.3.2",
    "tinyglobby": "catalog:",
    "tinypool": "^2.0.0",
    "tinyrainbow": "catalog:",
    "vite": "^6.0.0 || ^7.0.0",
    "why-is-node-running": "^2.3.0"
  },
  "devDependencies": {
    "@bomb.sh/tab": "^0.0.6",
    "@antfu/install-pkg": "^1.1.0",
    "@edge-runtime/vm": "^5.0.0",
    "@jridgewell/trace-mapping": "catalog:",
    "@sinonjs/fake-timers": "14.0.0",
    "@types/debug": "catalog:",
    "@types/estree": "catalog:",
    "@types/istanbul-lib-coverage": "catalog:",
    "@types/istanbul-reports": "catalog:",
    "@types/jsdom": "^21.1.7",
    "@types/node": "^22.18.6",
    "@types/picomatch": "^4.0.2",
    "@types/prompts": "^2.4.9",
    "@types/sinonjs__fake-timers": "^8.1.5",
    "acorn-walk": "catalog:",
    "birpc": "catalog:",
    "cac": "catalog:",
    "empathic": "^2.0.0",
    "flatted": "catalog:",
    "happy-dom": "^18.0.1",
    "jsdom": "^26.1.0",
    "local-pkg": "^1.1.2",
    "mime": "^4.1.0",
    "prompts": "^2.4.2",
    "strip-literal": "catalog:",
    "ws": "catalog:"
  }
}
