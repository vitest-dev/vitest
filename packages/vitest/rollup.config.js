import fs from 'node:fs'
import { builtinModules, createRequire } from 'node:module'
import { fileURLToPath } from 'node:url'
import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import nodeResolve from '@rollup/plugin-node-resolve'
import fg from 'fast-glob'
import { dirname, join, normalize, resolve } from 'pathe'
import { defineConfig } from 'rollup'
import dts from 'rollup-plugin-dts'
import esbuild from 'rollup-plugin-esbuild'
import license from 'rollup-plugin-license'
import c from 'tinyrainbow'

const require = createRequire(import.meta.url)
const pkg = require('./package.json')

const entries = {
  'path': 'src/paths.ts',
  'index': 'src/public/index.ts',
  'cli': 'src/node/cli.ts',
  'node': 'src/public/node.ts',
  'suite': 'src/public/suite.ts',
  'browser': 'src/public/browser.ts',
  'runners': 'src/public/runners.ts',
  'environments': 'src/public/environments.ts',
  'mocker': 'src/public/mocker.ts',
  'spy': 'src/integrations/spy.ts',
  'coverage': 'src/public/coverage.ts',
  'utils': 'src/public/utils.ts',
  'execute': 'src/public/execute.ts',
  'reporters': 'src/public/reporters.ts',
  // TODO: advanced docs
  'workers': 'src/public/workers.ts',

  // for performance reasons we bundle them separately so we don't import everything at once
  'worker': 'src/runtime/worker.ts',
  'workers/forks': 'src/runtime/workers/forks.ts',
  'workers/threads': 'src/runtime/workers/threads.ts',
  'workers/vmThreads': 'src/runtime/workers/vmThreads.ts',
  'workers/vmForks': 'src/runtime/workers/vmForks.ts',

  'workers/runVmTests': 'src/runtime/runVmTests.ts',

  'snapshot': 'src/public/snapshot.ts',
}

const dtsEntries = {
  index: 'src/public/index.ts',
  node: 'src/public/node.ts',
  environments: 'src/public/environments.ts',
  browser: 'src/public/browser.ts',
  runners: 'src/public/runners.ts',
  suite: 'src/public/suite.ts',
  config: 'src/public/config.ts',
  coverage: 'src/public/coverage.ts',
  utils: 'src/public/utils.ts',
  execute: 'src/public/execute.ts',
  reporters: 'src/public/reporters.ts',
  mocker: 'src/public/mocker.ts',
  workers: 'src/public/workers.ts',
  snapshot: 'src/public/snapshot.ts',
}

const external = [
  ...builtinModules,
  ...Object.keys(pkg.dependencies),
  ...Object.keys(pkg.peerDependencies),
  'worker_threads',
  'node:worker_threads',
  'node:fs',
  'node:os',
  'node:stream',
  'node:vm',
  'inspector',
  'vite-node/source-map',
  'vite-node/client',
  'vite-node/server',
  'vite-node/constants',
  'vite-node/utils',
  '@vitest/mocker',
  '@vitest/mocker/node',
  '@vitest/utils/diff',
  '@vitest/utils/ast',
  '@vitest/utils/error',
  '@vitest/utils/source-map',
  '@vitest/runner/utils',
  '@vitest/runner/types',
  '@vitest/snapshot/environment',
  '@vitest/snapshot/manager',
]

const dir = dirname(fileURLToPath(import.meta.url))

const plugins = [
  nodeResolve({
    preferBuiltins: true,
  }),
  json(),
  commonjs(),
  esbuild({
    target: 'node18',
  }),
]

export default ({ watch }) =>
  defineConfig([
    {
      input: entries,
      treeshake: true,
      output: {
        dir: 'dist',
        format: 'esm',
        chunkFileNames: 'chunks/[name].[hash].js',
      },
      external,
      plugins: [...plugins, !watch && licensePlugin()],
      onwarn,
    },
    {
      input: 'src/public/config.ts',
      output: [
        {
          file: 'dist/config.cjs',
          format: 'cjs',
        },
        {
          file: 'dist/config.js',
          format: 'esm',
        },
      ],
      external,
      plugins,
    },
    {
      input: dtsEntries,
      output: {
        dir: 'dist',
        entryFileNames: chunk =>
          `${normalize(chunk.name).replace('src/', '')}.d.ts`,
        format: 'esm',
        chunkFileNames: 'chunks/[name].[hash].d.ts',
      },
      external,
      plugins: [dts({ respectExternal: true })],
    },
  ])

function licensePlugin() {
  return license({
    thirdParty(dependencies) {
      // https://github.com/rollup/rollup/blob/master/build-plugins/generate-license-file.js
      // MIT Licensed https://github.com/rollup/rollup/blob/master/LICENSE-CORE.md
      const coreLicense = fs.readFileSync(resolve(dir, '../../LICENSE'))
      const licenses = new Set()
      const dependencyLicenseTexts = dependencies
        .filter(({ name }) => !name?.startsWith('@vitest/'))
        .sort(({ name: nameA }, { name: nameB }) =>
          nameA > nameB ? 1 : nameB > nameA ? -1 : 0,
        )
        .map(
          ({
            name,
            license,
            licenseText,
            author,
            maintainers,
            contributors,
            repository,
          }) => {
            let text = `## ${name}\n`
            if (license) {
              text += `License: ${license}\n`
            }

            const names = new Set()
            if (author && author.name) {
              names.add(author.name)
            }

            for (const person of maintainers.concat(contributors)) {
              if (person && person.name) {
                names.add(person.name)
              }
            }
            if (names.size > 0) {
              text += `By: ${Array.from(names).join(', ')}\n`
            }

            if (repository) {
              text += `Repository: ${repository.url || repository}\n`
            }

            if (!licenseText) {
              try {
                const pkgDir = dirname(
                  resolve(join(name, 'package.json'), {
                    preserveSymlinks: false,
                  }),
                )
                const [licenseFile] = fg.sync(`${pkgDir}/LICENSE*`, {
                  caseSensitiveMatch: false,
                })
                if (licenseFile) {
                  licenseText = fs.readFileSync(licenseFile, 'utf-8')
                }
              }
              catch {}
            }
            if (licenseText) {
              text += `\n${licenseText
                .trim()
                .replace(/(\r\n|\r)/g, '\n')
                .split('\n')
                .map(line => (line ? `> ${line}` : '>'))
                .join('\n')}\n`
            }
            licenses.add(license)
            return text
          },
        )
        .join('\n---------------------------------------\n\n')
      const licenseText
        = '# Vitest core license\n'
        + `Vitest is released under the MIT license:\n\n${coreLicense}\n# Licenses of bundled dependencies\n`
        + 'The published Vitest artifact additionally contains code with the following licenses:\n'
        + `${sortLicenses(licenses).join(', ')}\n\n`
        + `# Bundled dependencies:\n${dependencyLicenseTexts}`
      const existingLicenseText = fs.readFileSync('LICENSE.md', 'utf8')
      if (existingLicenseText !== licenseText) {
        fs.writeFileSync('LICENSE.md', licenseText)
        console.warn(
          c.yellow(
            '\nLICENSE.md updated. You should commit the updated file.\n',
          ),
        )
      }
    },
  })
}

function onwarn(message) {
  if (['EMPTY_BUNDLE', 'CIRCULAR_DEPENDENCY'].includes(message.code)) {
    return
  }
  console.error(message)
}

function sortLicenses(licenses) {
  let withParenthesis = []
  let noParenthesis = []
  licenses.forEach((license) => {
    if (/^\(/.test(license)) {
      withParenthesis.push(license)
    }
    else {
      noParenthesis.push(license)
    }
  })
  withParenthesis = withParenthesis.sort()
  noParenthesis = noParenthesis.sort()
  return [...noParenthesis, ...withParenthesis]
}
