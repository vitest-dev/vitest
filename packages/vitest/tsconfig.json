{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "isolatedDeclarations": true
  },
  "include": ["src/**/*"],
  "exclude": ["**/dist/**"]
}
