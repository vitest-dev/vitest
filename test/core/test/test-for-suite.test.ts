import { describe, expect, expectTypeOf, test } from 'vitest'

describe.for(['case1', 'case2'])(
  'basic %s',
  (...args) => {
    test('test', () => {
      expectTypeOf(args).toEqualTypeOf<[string]>()
      expect({ args }).matchSnapshot()
    })
  },
)

describe.for`
  a         | b
  ${'x'}    | ${true}
  ${'y'}    | ${false}
`(
  'template $a $b',
  (...args) => {
    test('test', () => {
      expectTypeOf(args).toEqualTypeOf<any[]>()
      expect({ args }).toMatchSnapshot()
    })
  },
)

describe.for([
  [1, 1],
  [1, 2],
  [2, 1],
])('add(%i, %i)', ([a, b]) => {
  test('test', () => {
    expect(a + b).matchSnapshot()
  })
})
