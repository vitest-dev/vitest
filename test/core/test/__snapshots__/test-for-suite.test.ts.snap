// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add(1, 1) > test 1`] = `2`;

exports[`add(1, 2) > test 1`] = `3`;

exports[`add(2, 1) > test 1`] = `3`;

exports[`basic case1 > test 1`] = `
{
  "args": [
    "case1",
  ],
}
`;

exports[`basic case2 > test 1`] = `
{
  "args": [
    "case2",
  ],
}
`;

exports[`template 'x' true > test 1`] = `
{
  "args": [
    {
      "a": "x",
      "b": true,
    },
  ],
}
`;

exports[`template 'y' false > test 1`] = `
{
  "args": [
    {
      "a": "y",
      "b": false,
    },
  ],
}
`;
